tests/trainer/test_trainer_cli_new.py

```diff
@@ -0,0 +1,245 @@
+
+import inspect
+import pickle
+import sys
+from argparse import Namespace
+from unittest import mock
+
+import pytest
+import torch
+
+import tests.base.develop_utils as tutils
+from pytorch_lightning import Trainer, LightningModule
+from pytorch_lightning.utilities.jsonargparse_utils import LightningArgumentParser, SaveConfigCallback, trainer_cli
+
+
+@mock.patch('argparse.ArgumentParser.parse_args')
+def test_default_args(mock_argparse, tmpdir):
+    """Tests default argument parser for Trainer"""
+    mock_argparse.return_value = Namespace(**Trainer.default_attributes())
+
+    # logger file to get meta
+    logger = tutils.get_default_logger(tmpdir)
+
+    parser = LightningArgumentParser(add_help=False)
+    args = parser.parse_args([])
+    args.logger = logger
+
+    args.max_epochs = 5
+    trainer = Trainer.from_argparse_args(args)
+
+    assert isinstance(trainer, Trainer)
+    assert trainer.max_epochs == 5
+
+
+@pytest.mark.parametrize('cli_args', [
+    ['--accumulate_grad_batches=22'],
+    ['--weights_save_path=./'],
+    []
+])
+def test_add_argparse_args_redefined(cli_args):
+    """Redefines some default Trainer arguments via the cli and
+    tests the Trainer initialization correctness.
+    """
+    parser = LightningArgumentParser(add_help=False)
+    parser.add_trainer_args(Trainer, None)
+
+    args = parser.parse_args(cli_args)
+
+    # make sure we can pickle args
+    pickle.dumps(args)
+
+    # Check few deprecated args are not in namespace:
+    for depr_name in ('gradient_clip', 'nb_gpu_nodes', 'max_nb_epochs'):
+        assert depr_name not in args
+
+    trainer = Trainer.from_argparse_args(args=args)
+    pickle.dumps(trainer)
+
+    assert isinstance(trainer, Trainer)
+
+
+@pytest.mark.parametrize('cli_args', [
+    ['--callbacks=1', '--logger'],
+    ['--foo', '--bar=1']
+])
+def test_add_argparse_args_redefined_error(cli_args, monkeypatch):
+    """Asserts that an error raised in case of passing not default cli arguments."""
+
+    class _UnkArgError(Exception):
+        pass
+
+    def _raise():
+        raise _UnkArgError
+
+    parser = LightningArgumentParser(add_help=False)
+    parser.add_trainer_args(Trainer, None)
+
+    monkeypatch.setattr(parser, 'exit', lambda *args: _raise(), raising=True)
+
+    with pytest.raises(_UnkArgError):
+        parser.parse_args(cli_args)
+
+
+@pytest.mark.parametrize(['cli_args', 'expected'], [
+    #pytest.param('--auto_lr_find --auto_scale_batch_size power',
+    #             {'auto_lr_find': True, 'auto_scale_batch_size': 'power'}),
+    #pytest.param('--auto_lr_find any_string --auto_scale_batch_size',
+    #             {'auto_lr_find': 'any_string', 'auto_scale_batch_size': True}),
+    #pytest.param('--auto_lr_find t --auto_scale_batch_size ON',
+    #             {'auto_lr_find': True, 'auto_scale_batch_size': True}),
+    #pytest.param('--auto_lr_find 0 --auto_scale_batch_size n',
+    #             {'auto_lr_find': False, 'auto_scale_batch_size': False}),
```

Comment:
these do not work?

Reply:
This is explained in the description. They don't work, but maybe they shouldn't. These are booleans, so their values should be true or false.

Reply:
Removed the commented params and added a few more since according to the yaml standard `On`, `Off`, `Yes` and `No` also work.